interfaces/builtin/gpio.go

```diff
@@ -0,0 +1,164 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var gpioSysfsGpioBase = "/sys/class/gpio/gpio"
+var gpioSysfsExport = "/sys/class/gpio/export"
+
+// GpioInterface type
+type GpioInterface struct{}
+
+// String returns the same value as Name().
+func (iface *GpioInterface) String() string {
+	return iface.Name()
+}
+
+// Name of the GpioInterface
+func (iface *GpioInterface) Name() string {
+	return "gpio"
+}
+
+// SanitizeSlot checks the slot definition is valid
+func (iface *GpioInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Paranoid check this right interface type
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// We will only allow creation of this type of slot by a gadget or OS snap
+	if !(slot.Snap.Type == "gadget" || slot.Snap.Type == "os") {
+		return fmt.Errorf("gpio slots only allowed on gadget or os snaps")
+	}
+
+	// Must have a GPIO number
+	number, ok := slot.Attrs["number"]
+	if !ok {
+		return fmt.Errorf("gpio slot must have a number attribute")
+	}
+
+	// Valid values of number
+	switch t := number.(type) {
+	case int:
+		break
+	default:
+		_ = t
+		return fmt.Errorf("gpio slot number attribute must be an int")
+	}
+
+	// Must have a direction
+	direction, ok := slot.Attrs["direction"].(string)
```

Comment:
I'm not quite sure about this attribute. This effectively splits the gpio interface in two. We can never connect something expecting input with something that can only read it. We'll need to talk about this.